atc/db/resource.go

```diff
@@ -953,3 +956,243 @@ func requestScheduleForJobsUsingResource(tx Tx, resourceID int) error {
 
 	return nil
 }
+
+var (
+	downStreamCausalityQuery = `
+WITH RECURSIVE build_ids AS (
+		SELECT DISTINCT i.build_id
+			FROM build_resource_config_version_inputs i
+			WHERE i.resource_id=$1 AND i.version_md5=$2
+	UNION ALL
+		SELECT DISTINCT bp.to_build_id AS build_id
+		FROM build_ids bi
+		INNER JOIN build_pipes bp ON bi.build_id = bp.from_build_id
+		INNER JOIN build_resource_config_version_inputs i ON i.build_id = bi.build_id
```

Comment:
I ran this query against `hush-house` on a resource version that had approx 200 builds, and it seemed to take around 500 ms to run https://explain.depesz.com/s/4Acp. I think a chunk of time is spent on this join onto `build_resource_config_version_inputs` in order to remove any builds that have the `resource_id` as an input. Just wanted to mention it in case we can think of anything to refactor this query in order to not need us to do that where clause!